sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/AsyncCacheNonBlocking.java

```diff
@@ -0,0 +1,208 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.cosmos.implementation.caches;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Mono;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Function;
+
+public class AsyncCacheNonBlocking<TKey, TValue> {
+    private final Logger logger = LoggerFactory.getLogger(AsyncCacheNonBlocking.class);
+    private final ConcurrentHashMap<TKey, AsyncLazyWithRefresh<TValue>> values;
+    private final IEqualityComparer<TKey> keyEqualityComparer;
+
+    public AsyncCacheNonBlocking(ConcurrentHashMap<TKey, AsyncLazyWithRefresh<TValue>> values, IEqualityComparer<TKey> keyEqualityComparer) {
+        this.values = values;
+        this.keyEqualityComparer = keyEqualityComparer;
+    }
+
+    public AsyncCacheNonBlocking(IEqualityComparer<TKey> keyEqualityComparer) {
+        this(new ConcurrentHashMap<>(), keyEqualityComparer);
+    }
+
+    public AsyncCacheNonBlocking() {
+        this((value1, value2) -> {
+            if (value1 == value2)
+                return true;
+            if (value1 == null || value2 == null)
+                return false;
+            return value1.equals(value2);
+        });
+    }
+
+    /**
+     *
+     * <p>
+     * If another initialization function is already running, new initialization function will not be started.
+     * The result will be result of currently running initialization function.
+     * </p>
+     *
+     * <p>
+     * If previous initialization function is successfully completed it will return the value. It is possible this
+     * value is stale and will only be updated after the force refresh task is complete.
+     * </p>
+     *
+     * <p>
+     * If previous initialization function failed - new one will be launched.
+     * </p>
+     *
+     * @param key Key for which to get a value.
+     * @param singleValueInitFunc Initialization function.
+     * @param forceRefresh Force refresh for refreshing the cache
+     * @return Cached value or value returned by initialization function.
+     */
+    public Mono<TValue> getAsync(
+        TKey key,
+        Function<TValue, Mono<TValue>> singleValueInitFunc,
+        boolean forceRefresh) {
+
+        AsyncLazyWithRefresh<TValue> initialLazyValue = values.get(key);
+        if (initialLazyValue != null) {
+            logger.debug("cache[{}] exists", key);
+            return initialLazyValue.getValueAsync().flatMap(value -> {
+                if(!forceRefresh) {
```

Comment:
Previously, we also allowed the obsoleteValue checking. Do we purposely removed it here? 

```
public Mono<TValue> getAsync(
            TKey key,
            TValue obsoleteValue,
            Callable<Mono<TValue>> singleValueInitFunc) {
```

Reply:
We are not using obsoleteValue here because we have the forceRefresh which when true will refresh the cache. So, having the obsoleteValue check will be redundant.